moved/src/move_execution/state.rs

```diff
@@ -0,0 +1,360 @@
+use {
+    crate::{
+        move_execution::{quick_get_eth_balance, quick_get_nonce},
+        primitives::U256,
+    },
+    move_binary_format::errors::PartialVMError,
+    move_core_types::{
+        account_address::AccountAddress, effects::ChangeSet, resolver::MoveResolver,
+    },
+    move_table_extension::TableResolver,
+    move_vm_test_utils::InMemoryStorage,
+};
+
+pub type Balance = U256;
+pub type Nonce = u64;
+pub type BlockHeight = u64;
+
+pub trait StateQueries {
+    /// The associated storage type for querying the blockchain state.
+    type Storage;
+
+    /// Queries the blockchain state version corresponding with block `height` for amount of base
+    /// token associated with `account`.
+    fn balance_at(&self, account: AccountAddress, height: BlockHeight) -> Balance;
+
+    /// Queries the blockchain state version corresponding with block `height` for the nonce value
+    /// associated with `account`.
+    fn nonce_at(&self, account: AccountAddress, height: BlockHeight) -> Nonce;
+}
+
+#[derive(Debug)]
+pub struct StateMemory {
+    changes: Vec<ChangeSet>,
+}
+
+impl Default for StateMemory {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl StateMemory {
+    pub fn from_genesis(changes: ChangeSet) -> Self {
+        Self {
+            changes: vec![changes],
+        }
+    }
+
+    pub fn new() -> Self {
+        Self {
+            changes: Vec::new(),
+        }
+    }
+
+    pub fn add(&mut self, change: ChangeSet) {
+        self.changes.push(change);
+    }
+
+    pub fn resolver(&self, height: usize) -> impl MoveResolver<PartialVMError> + TableResolver {
+        let mut state = InMemoryStorage::new();
+
+        if let Some(last_index) = self.changes.len().checked_sub(1) {
+            let height = height.min(last_index);
+
+            for change in self.changes[0..=height].iter() {
```

Comment:
The amount of work here is linear in the height of the request, which will be totally unacceptable after more than a few thousand blocks. I think if our strategy is to store change sets then we also need to store occasional complete state checkpoints to avoid starting from 0 on each query.

I still think the better approach (for now) is to not store change sets at all and instead use the merkle store for history. The idea is that we are already storing the complete set of merkle trie nodes which are part of the merkle tries of all past blocks. To look up a state element at height H we look up the block at H, get its state root R and then index into the merkle trie with root R to find the leaf node we need. This make the amount of work independent of block height. The trade off is that queries will be sensitive to the depth of the trie, but this should grow much slower than H so that it will be fine in the medium term. And avoiding direct merkle queries something we can address in the future with a more clever flat storage implementation.